File: Swift/프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.swift
import Foundation

func solution(_ n:Int, _ words:[String]) -> [Int] {
    var turnCount: [Int: Int] = [:] //index : count
    var wordSet: Set<String> = Set()
    
    var lastChar: String = String(words.first![words.first!.startIndex])
    var result: [Int] = [0, 0]
    for wordIndex in 0..<words.count {
        let person: Int = (wordIndex % n) + 1
        var count = (turnCount[person] ?? 0) + 1
        turnCount[person] = count
        let word = words[wordIndex]
        guard lastChar == String(word[word.startIndex]), !wordSet.contains(word) else { 
            result[0] = person
            result[1] = count
            return result
        }
        lastChar = String(word[word.index(before:word.endIndex)])
        wordSet.insert(word)
    }
    
    return result
}

File: Java/프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.HashSet;
import java.util.HashMap;

class Solution {
    public int[] solution(int n, String[] words) {
        HashSet<String> wordSet = new HashSet<>();
        HashMap<Integer, Integer> turnCount = new HashMap<>();
        
        char lastChar = words[0].charAt(0);
        int[] result = new int[2];
        for (int i = 0; i<words.length; i++) {
            int person = (i % n) + 1;
            int count = turnCount.getOrDefault(person, 0) + 1;
            turnCount.put(person, count);
            String word = words[i];
            if (lastChar != word.charAt(0) || wordSet.contains(word)) {
                result[0] = person;
                result[1] = count;
                return result;
            }
            
            lastChar = word.charAt(word.length()-1);
            wordSet.add(word);
            
        }

        return result;
    }
}